File: src/testing/testing.ts
// `cappu test` (nikeee/cappu#16): compile src/test/java against the main
// classes plus lib/classes and lib/test-classes, then run the JUnit Platform
// Console Launcher over the result. Self-contained, mirroring src/packages/:
// nothing here prints and `exec` is injectable; the CLI streams the actual
// JUnit run (live output is the point of a test runner).

import { spawnSync } from "node:child_process";
import { accessSync, constants, existsSync, mkdirSync, realpathSync, writeFileSync } from "node:fs";
import { delimiter, dirname, join } from "node:path";

import { type CompileDiagnostic, parseJavacDiagnostics } from "../compiler/javacDiagnostics.ts";
import { expandedJarDirs } from "../compiler/javacPaths.ts";
import {
  type CappuConfig,
  DEFAULT_TEST_CLASS_PATH,
  DEFAULT_TEST_RESOURCE_PATH,
  DEFAULT_TEST_SOURCE_PATH,
  resolveConfigPath,
} from "../config.ts";
import { configuredSources, storePathFor } from "../install.ts";
import { provisionedJava, provisionedJavac } from "../jdks/index.ts";
import type { Coordinates, PackageSource } from "../packages/index.ts";
import { findJavaFiles } from "../workspace.ts";

// Derived build state, like .cappu/generated-sources (gitignored via /.cappu/).
export const TEST_BUILD_ROOT = ".cappu/test-build";

export function mainClassesDir(config: CappuConfig): string {
  return resolveConfigPath(config, join(TEST_BUILD_ROOT, "classes"));
}

export function testClassesDir(config: CappuConfig): string {
  return resolveConfigPath(config, join(TEST_BUILD_ROOT, "test-classes"));
}

/** All .java files under src/test/java (not configurable yet). */
export function findTestSources(config: CappuConfig): string[] {
  return findJavaFiles(resolveConfigPath(config, DEFAULT_TEST_SOURCE_PATH));
}

// Main classes + compile deps + test deps: what test sources compile against
// and (plus the compiled tests and test resources) what they run with.
function dependencyClassPath(config: CappuConfig): string[] {
  return [
    mainClassesDir(config),
    ...expandedJarDirs(config.compilerOptions.classPath.map(p => resolveConfigPath(config, p))),
    ...expandedJarDirs([resolveConfigPath(config, DEFAULT_TEST_CLASS_PATH)]),
  ];
}

/** The classpath the JUnit launcher runs with. */
export function testRuntimeClassPath(config: CappuConfig): string[] {
  const testResources = resolveConfigPath(config, DEFAULT_TEST_RESOURCE_PATH);
  return [
    testClassesDir(config),
    ...dependencyClassPath(config),
    ...(existsSync(testResources) ? [testResources] : []),
  ];
}

export interface ExecResult {
  status: number | null;
  stderr: string;
  error?: Error;
}

export type Exec = (bin: string, args: string[]) => ExecResult;

const defaultExec: Exec = (bin, args) => {
  const result = spawnSync(bin, args, { stdio: ["ignore", "ignore", "pipe"] });
  return { status: result.status, stderr: result.stderr?.toString() ?? "", error: result.error };
};

/** The javac arguments compiling the test sources. */
export function compileTestsArgs(config: CappuConfig, sources: readonly string[]): string[] {
  return [
    "-d",
    testClassesDir(config),
    "-encoding",
    "UTF-8",
    "-cp",
    dependencyClassPath(config).join(delimiter),
    ...sources,
  ];
}

/** Compile src/test/java; diagnostics non-empty on failure. */
export function compileTests(
  config: CappuConfig,
  sources: readonly string[],
  exec: Exec = defaultExec,
): CompileDiagnostic[] {
  mkdirSync(testClassesDir(config), { recursive: true });
  const javac = provisionedJavac(config) ?? config.compilerOptions.javac;
  const result = exec(javac, compileTestsArgs(config, sources));
  if (result.error || result.status === null) {
    return [
      {
        severity: "error",
        message: `compiling tests needs javac: '${javac}' could not run (${result.error?.message ?? "unknown error"})`,
      },
    ];
  }
  if (result.status !== 0) {
    const diagnostics = parseJavacDiagnostics(result.stderr);
    return diagnostics.length > 0
      ? diagnostics
      : [
          {
            severity: "error",
            message: `test compilation failed: ${result.stderr.trim().slice(-400)}`,
          },
        ];
  }
  return [];
}

// The JUnit Platform Console Launcher: a TOOL, not a project dependency - it
// never appears in cappu.json or the lockfile. Pinned; bundles the platform
// and the jupiter/vintage engines, so projects only declare junit-jupiter.
export const CONSOLE_LAUNCHER: Coordinates = {
  groupId: "org.junit.platform",
  artifactId: "junit-platform-console-standalone",
  version: "1.12.2",
};

/**
 * The launcher jar's path in the global package store, downloading it there
 * on first use. Sources are injectable for tests.
 */
export async function consoleLauncherJar(
  config: CappuConfig,
  sources: readonly PackageSource[] = configuredSources(config),
): Promise<string> {
  const path = storePathFor(CONSOLE_LAUNCHER);
  if (!path) throw new Error("unreachable: launcher coordinates are store-safe");
  if (existsSync(path)) return path;
  for (const source of sources) {
    const bytes = await source.getArtifact?.(CONSOLE_LAUNCHER);
    if (!bytes) continue;
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, bytes);
    return path;
  }
  throw new Error(
    `could not download ${CONSOLE_LAUNCHER.groupId}:${CONSOLE_LAUNCHER.artifactId}:${CONSOLE_LAUNCHER.version} from any package source`,
  );
}

/** The `java` arguments running the launcher over the compiled tests. */
export function testRunArgs(config: CappuConfig, launcherJar: string): string[] {
  return [
    "-jar",
    launcherJar,
    "execute",
    "--class-path",
    testRuntimeClassPath(config).join(delimiter),
    "--scan-class-path",
  ];
}

/**
 * The java launcher tests run under: the provisioned JDK's, else the sibling
 * of the resolved javac (so a PATH skew between javac and java cannot produce
 * UnsupportedClassVersionError), else plain "java" from PATH.
 */
export function resolveJava(config: CappuConfig): string {
  const provisioned = provisionedJava(config);
  if (provisioned) return provisioned;
  const javac = provisionedJavac(config) ?? config.compilerOptions.javac;
  try {
    const real = realpathSync(javac.includes("/") || javac.includes("\\") ? javac : onPath(javac));
    const sibling = join(dirname(real), process.platform === "win32" ? "java.exe" : "java");
    accessSync(sibling, constants.X_OK);
    return sibling;
  } catch {
    return "java";
  }
}

// Resolve a bare binary name against PATH (throws when absent; the caller
// falls back).
function onPath(name: string): string {
  for (const dir of (process.env.PATH ?? "").split(delimiter)) {
    if (!dir) continue;
    const candidate = join(dir, name);
    try {
      accessSync(candidate, constants.X_OK);
      return candidate;
    } catch {
      // keep looking
    }
  }
  throw new Error(`${name} not on PATH`);
}

File: togo/internal/testing/testing.go
// Package testing implements `cappu test` (nikeee/cappu#16): compile
// src/test/java against the main classes plus the lib/test-classes, then run the
// JUnit Console Launcher over the result. Print-free; the CLI streams the JUnit
// run. Port of src/testing/.
package testing

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/nikeee/cappu/internal/build"
	"github.com/nikeee/cappu/internal/config"
	"github.com/nikeee/cappu/internal/install"
	"github.com/nikeee/cappu/internal/javacdiag"
	"github.com/nikeee/cappu/internal/jdks"
	"github.com/nikeee/cappu/internal/packages"
	"github.com/nikeee/cappu/internal/sources"
)

// testBuildRoot is derived build state under .cappu (gitignored via /.cappu/).
const testBuildRoot = ".cappu/test-build"

// MainClassesDir is where the project's main classes compile to for a test run.
func MainClassesDir(cfg *config.Config) string {
	return cfg.ResolvePath(filepath.Join(testBuildRoot, "classes"))
}

// TestClassesDir is where the compiled test classes go.
func TestClassesDir(cfg *config.Config) string {
	return cfg.ResolvePath(filepath.Join(testBuildRoot, "test-classes"))
}

// FindTestSources lists all .java files under src/test/java.
func FindTestSources(cfg *config.Config) []string {
	return build.JavaFilesIn(cfg.ResolvePath(config.DefaultTestSourcePath))
}

// dependencyClassPath is what test sources compile against: the main classes,
// the compile deps and the test deps (jar-expanded).
func dependencyClassPath(cfg *config.Config) []string {
	cp := []string{MainClassesDir(cfg)}
	cp = append(cp, build.ClassPath(cfg)...)
	cp = append(cp, build.ExpandJarDirs([]string{cfg.ResolvePath(config.DefaultTestClassPath)})...)
	return cp
}

// TestRuntimeClassPath is the classpath the JUnit launcher runs with.
func TestRuntimeClassPath(cfg *config.Config) []string {
	cp := []string{TestClassesDir(cfg)}
	cp = append(cp, dependencyClassPath(cfg)...)
	testResources := cfg.ResolvePath(config.DefaultTestResourcePath)
	if _, err := os.Stat(testResources); err == nil {
		cp = append(cp, testResources)
	}
	return cp
}

// compileTestsArgs builds the javac arguments for the test compile. Port of
// compileTestsArgs.
func compileTestsArgs(cfg *config.Config, sources []string) []string {
	args := []string{"-d", TestClassesDir(cfg), "-encoding", "UTF-8"}
	if cfg.CompilerOptions.Release != nil {
		args = append(args, "--release", strconv.Itoa(*cfg.CompilerOptions.Release))
	}
	args = append(args, "-cp", strings.Join(dependencyClassPath(cfg), string(os.PathListSeparator)))
	return append(args, sources...)
}

// CompileTests compiles src/test/java; the returned diagnostics are non-empty on
// failure (empty on success). The test-classes dir is wiped first so a
// since-deleted test cannot still be discovered by --scan-class-path. Port of
// compileTests.
func CompileTests(cfg *config.Config, testSources []string) []javacdiag.CompileDiagnostic {
	dir := TestClassesDir(cfg)
	_ = os.RemoveAll(dir)
	_ = os.MkdirAll(dir, 0o755)
	javac := build.Javac(cfg)
	cmd := exec.Command(javac, compileTestsArgs(cfg, testSources)...)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	err := cmd.Run()
	if err == nil {
		return nil
	}
	var exit *exec.ExitError
	if !errors.As(err, &exit) {
		return []javacdiag.CompileDiagnostic{{Severity: "error",
			Message: fmt.Sprintf("compiling tests needs javac: '%s' could not run (%s)", javac, err)}}
	}
	diagnostics := javacdiag.ParseJavacDiagnostics(stderr.String())
	if len(diagnostics) == 0 {
		s := strings.TrimSpace(stderr.String())
		if len(s) > 400 {
			s = s[len(s)-400:]
		}
		diagnostics = []javacdiag.CompileDiagnostic{{Severity: "error", Message: "test compilation failed: " + s}}
	}
	return diagnostics
}

// consoleLauncher is the pinned JUnit Platform Console Launcher: a TOOL, never a
// project dependency, so it never appears in cappu.json or the lockfile.
var consoleLauncher = packages.NewCoordinates("org.junit.platform", "junit-platform-console-standalone", "1.12.2")

// ConsoleLauncherJar returns the launcher jar's path in the global store,
// downloading it there on first use.
func ConsoleLauncherJar(cfg *config.Config) (string, error) {
	path, ok := install.StorePathFor(consoleLauncher)
	if !ok {
		return "", fmt.Errorf("unreachable: launcher coordinates are store-safe")
	}
	if _, err := os.Stat(path); err == nil {
		return path, nil
	}
	for _, source := range sources.Configured(cfg) {
		bytes, err := source.GetArtifact(consoleLauncher)
		if err != nil {
			return "", err
		}
		if bytes == nil {
			continue
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return "", err
		}
		if err := os.WriteFile(path, bytes, 0o644); err != nil {
			return "", err
		}
		return path, nil
	}
	return "", fmt.Errorf("could not download %s from any package source", consoleLauncher.String())
}

// TestRunArgs are the java arguments running the launcher over the tests.
func TestRunArgs(cfg *config.Config, launcherJar string) []string {
	return []string{
		"-jar", launcherJar, "execute",
		"--class-path", strings.Join(TestRuntimeClassPath(cfg), string(os.PathListSeparator)),
		"--scan-class-path",
	}
}

// ResolveJava is the java launcher tests run under: the provisioned JDK's, else
// the sibling of the resolved javac (so a PATH skew between javac and java
// cannot cause UnsupportedClassVersionError), else plain "java". A bare javac
// name is looked up on PATH and symlink-resolved first, matching the TS
// resolveJava, so e.g. a /usr/bin/javac that points at JDK 25 picks that JDK's
// java rather than a different default `java` on PATH.
func ResolveJava(cfg *config.Config) string {
	if java := jdks.ProvisionedJava(cfg); java != "" {
		return java
	}
	javac := build.Javac(cfg)
	path := javac
	if !strings.ContainsAny(javac, `/\`) {
		if p, err := exec.LookPath(javac); err == nil {
			path = p
		}
	}
	if real, err := filepath.EvalSymlinks(path); err == nil {
		path = real
	}
	name := "java"
	if runtime.GOOS == "windows" {
		name = "java.exe"
	}
	sibling := filepath.Join(filepath.Dir(path), name)
	if _, err := os.Stat(sibling); err == nil {
		return sibling
	}
	return "java"
}
